simulator/src/main/java/com/hedera/block/simulator/grpc/PublishStreamGrpcClientImpl.java

```diff
@@ -72,12 +83,28 @@ public boolean streamBlockItem(BlockItem blockItem) {
      */
     @Override
     public boolean streamBlock(Block block) {
+        List<com.hedera.hapi.block.stream.protoc.BlockItem> blockItemsProtoc = new ArrayList<>();
         for (BlockItem blockItem : block.items()) {
-            streamBlockItem(blockItem);
+            blockItemsProtoc.add(Translator.fromPbj(blockItem));
         }
 
-        // wait for ack on the block
-        // if and when the ack is received return true
+        final int blockItemsNumberOfBatches =
+                block.items().size() / blockStreamConfig.blockItemsBatchSize();
```

Comment:
I have a concern here with the way this is being batched, hope I am wrong. This int/int division should result in a whole int that would be the closest to the next int (if result is 2.3 then 2 will be returned, if result is 2.9 then 3 will be returned). Please correct me if I am wrong, but this undeterministic behavior will result in bugs.

Also, lets extract this batching logic in the `common` module, my proposal is something of the sort (or propose similar) this will return the chunks and we can foreach them functionally:
```java
public final class ChunkUtils {
    public static <T> List<List<T>> chunkify(final Collection<T> collection, final int chunkSize) {
        Objects.requireNonNull(collection);
        if (chunkSize <= 0) {
            throw new IllegalArgumentException("Chunk size must be greater than 0");
        }
        return IntStream.iterate(0, i -> i < collection.size(), i -> i + chunkSize)
            .mapToObj(i -> List.copyOf(collection).subList(i, Math.min(i + chunkSize, collection.size())))
            .collect(Collectors.toList());
    }

    private ChunkUtils() {
    }
}
```

Reply:
I like your approach, but I rather not use `.forEach` for concurrency and performance concerns.  other than that, happy to implement this improvement, and YES! this is a bug waiting to happen, will use your approah. Thank you!! 🙏

Reply:
✅ Done!

Reply:
Noted, glad you like the idea!
For me is resolved, either with the advanced for or the `forEach`, up to you.

I wonder however what concerns do you have about performance and concurrency by using `forEach`? The implementation internally is a for loop again, (with tiny overhead probably against the advanced for which uses the iterator, but I am not sure if it even exists as an overhead)? And about the concurrency what issues could arise? Maybe there is something I do not know, please let me know I would love to learn! Otherwise the advanced for loop and the `forEach` look the same to me.

Reply:
Is my understanding that the `.forEach` is executed in parallel (could be wrong) , also, is less performant (this is true). But ultimately, a regular For gives you more flexibility overall and a cleaner and simpler code.